### ISSUE TITLE
dartaotruntime help shows --observe but it does not work

### ISSUE THREAD

**Issue by Mabsten** [dart-lang/sdk#44651]
`dart run` and `dartaotruntime` help menu show the `--observe` flag, but it doesn't work with dartaotruntime. Is the Observatory tool supported to inspect AOT compiled snapshot? (the [Observatory doc guide](https://dart-lang.github.io/observatory/) is not precise on this).

Steps to reproduce:
- Compile a simple pure-dart app with `dart2native main.dart --output-kind=aot`
- Run the snapshot with `dartaotruntime --observe main.aot`
- It shows this error:
  > Setting VM flags failed: Unrecognized flags: pause-isolates-on-exit, pause-isolates-on-unhandled-exceptions, profiler

To avoid the unrecognized implicit flags, I have also tried with 
`dartaotruntime --enable-vm-service --disable-service-auth-codes main.aot`
In this way no error appears, but no log appears with the Observatory tool address (and indeed at the default address `127.0.0.1:8181` the tool isn't available; I had compiled an app calling `stdin.readLineSync()`, so also without the pause-isolates-on-exit implicit flag the service should remain available).

Thanks for your attention.

**Comment by mraleph** [comment]
`dartaotruntime` is built in product mode and does not contain/support Observatory. We should update `--help` to reflect that. If it is profiling that you would like to do then on Linux you can use `perf` to profile an AOT compiled application. I am not sure what would be the best profile on Windows (things are a bit complicated because we emit ELF snapshots on Windows and then use custom ELF loader to load them - so I doubt any profiler would actually be able to pick up the symbols from the snapshot).

**Comment by maks** [comment]
@mraleph I'm sorry to resurrect this old issue and I have a vague feeling of deja vu that I've asked you this before (sorry if I have) but I can't seem to find any other answer to this, as according to [this table](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/feature_availability.md) which I found via #40344 a bunch of the vm service features are supposed to be available in "all modes" which I take to mean including AOT and I would like to access them in AOT exe's. 

So I wanted to ask if that table is still correct? and if it is, is it possible programmatically or via a commandline flag to the exec to enable the vm service to run in a aot exe? I'd prefer to connect to it via its usual websocket but if thats not possible, even being able to connect to the vm service "internally" with Dart code running in the same process would be useful for my use case (a httpd server app).

**Comment by mraleph** [comment]
@maks There is _no VM service at all_ in the product mode, so the table does not apply. It documents which features are supported assuming vm-service is actually present. In Flutter parlance `dart compile exe` produces `--release` builds. 

Which specific features of `vm-service` are you interested in?

**Comment by maks** [comment]
Thanks for the clarification @mraleph and taking a look at this! 👍🏻 

For my current project I'd like to be able to use from the vm services the following APIs for Isolates:

* https://pub.dev/documentation/vm_service/latest/vm_service/VM/isolates.html
* https://pub.dev/documentation/vm_service/latest/vm_service/VM/isolateGroups.html
* https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/number.html
* https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/startTime.html
* https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/runnable.html
* https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/libraries.html

and then also for mem usage:
https://pub.dev/documentation/vm_service/latest/vm_service/MemoryUsage-class.html

and (a nice to have) would be profiling:
https://pub.dev/documentation/vm_service/latest/vm_service/VmServiceInterface/getCpuSamples.html
https://pub.dev/documentation/vm_service/latest/vm_service/CpuSamples-class.html

I do have some hacky workarounds atm, eg. I'm currently working around lack of access to Isolate Id's & VM by abusing Isolate debug name and keeping my own hashmap for the mappings but that's less than ideal.

**Comment by mraleph** [comment]
> https://pub.dev/documentation/vm_service/latest/vm_service/VM/isolates.html
> https://pub.dev/documentation/vm_service/latest/vm_service/VM/isolateGroups.html
> https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/number.html
> https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/startTime.html
> https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/runnable.html

Could you share any specifics on how would you use the data these APIs expose?

> https://pub.dev/documentation/vm_service/latest/vm_service/Isolate/libraries.html

Just an example here: when you compile in product mode we throw away information about most of the libraries away, because it makes snapshot smaller and faster to start (and have smaller memory footprint).

> and (a nice to have) would be profiling:

My recommended way to profile AOT code is to use a native profiling tool (e.g. `perf`) rather than builtin sampling profiler.

**Comment by maks** [comment]
@mraleph sorry I should have included more details on my use case: I'm building an new backend framework in Dart, styled more towards Erlang/Elixir rather than styled after NodeJS as the existing shelf, dartfrog etc are.

To that end I intend to use large number of Isolates running workloads that are able to be modified if needed at runtime (eg. initially in scripting langs: Lua or JS, later on WASM engine running anything compiled to wasm bytecode (including Dart 🙂 ) so I would like to have runtime introspection for debugging and devops purposes.

Hence I'd like to be able to have access to listing a VM's Isolates at runtime, get some runtime metadata about them, etc. Would this information not be available, even if a AOT compiled exe/snapshot during runtime?

Likewise for mem usage data at runtime.

As I mentioned, the profiling is more a nice to have, so not worried about that not being available, especially since scripting or wasm runtimes already have facilities for this too.

Hope that helps explain what I'm after.

**Comment by maks** [comment]
@mraleph I've done some more thinking and work on this.

Firstly I've remembered that I can actually use the [existing `ProcessInfo` in dart:io](https://api.flutter.dev/flutter/dart-io/ProcessInfo/currentRss.html) for accessing memory usage, so apologies I don't need the vm service for this after all.

For the Isolate related API's I think the key thing I'm really after is just to be able to access the VM assigned unique ID for an Isolate, so I'll open a new issue specifically for just that request.

**Comment by maks** [comment]
@mraleph sorry to bug you again but I'm still plugging away at my project and now trying to figure out if there is a way I can use just Dart (if possible without calling out with FFI to OS specific libs or using OS specifics like `/proc`) to get Dart processes CPU usage information in AOT mode and I noticed this in [the Flutter docs:](https://docs.flutter.dev/perf/ui-performance#connect-to-a-physical-device)

>but profile and release builds are pre-compiled to native instructions (also called “ahead of time”, or AOT) before the app is loaded onto the device

which made me wonder, is there some mechanism available that Flutter profile mode uses to collect profiling data in AOT mode? I know I can dig through the Flutter engine src to try to find out, but I thought I'd ask here as I suspect you probably know the answer but also the means by which it works and I could or couldn't make use of it in just a Dart SDK AOT compiled app (vs a Flutter app).

**Comment by mraleph** [comment]
Flutter `--profile` builds are using Dart "AOT release" builds, which is an AOT build which includes vm-service. Flutter `--release` uses Dart "AOT product" build (no vm-service). When we were implementing `dart compile exe` we decided that we don't want to support the zoo of modes - so we decided that `dart compile exe` will use "AOT product" mode.

**Comment by maks** [comment]
Thanks again for the explanation @mraleph ! That is disappointing for my (admittedly very niche) purposes but a completely understandable decision in terms of Dart team resources and things to focus on and I am still very happy that Dart does support an AOT compile mode, being able to produce a stand along executable is very useful for me.

**Comment by YaroCabbage** [comment]
> Flutter `--profile` builds are using Dart "AOT release" builds, which is an AOT build which includes vm-service. Flutter `--release` uses Dart "AOT product" build (no vm-service). When we were implementing `dart compile exe` we decided that we don't want to support the zoo of modes - so we decided that `dart compile exe` will use "AOT product" mode.

@mraleph Hi! But is it possible to monitor that cpu and memory for each isolate using perf on linux? As I understand that's not an easy task even if using JIT compilation with developer package.

**Comment by mraleph** [comment]
@YaroCabbage any specific reason why you would prefer _per-isolate_ CPU usage breakdown? 

You can certainly monitor memory usage and CPU consumption in aggregate.

**Comment by YaroCabbage** [comment]
> @YaroCabbage any specific reason why you would prefer _per-isolate_ CPU usage breakdown?
> 
> You can certainly monitor memory usage and CPU consumption in aggregate.

Yes! I'm going to use dart for backend and I'd like to run heavy compilations for each user independently using isolates. Cpu/memory usage breakdown per-isolate will make possible to track how much of system resource user currently using without having overhead running separate instances of compiled app. Also there are other advantages: separating space for a little more security and better exception handling.

**Comment by davidmartos96** [comment]
@mraleph I'm in a similar situation as maks, I would like to use the Memory trace (#instances, ...) on a AOT Dart build.
The objective is to debug a memory leak on a Dart server, which unfortunately only manifests on production.

I understand that `dart compile exe` generates the AOT product mode, with no VM Service, but since Flutter is able to create a profile build, is there any kind of flag we could use, or a combination of commands (like dartaotruntime) to achieve that? Or is Flutter using Dart in a way that is not achievable through the CLI?
